Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1156 @@
+and then removed, in favor of RCU.
+
+Implementing RCU for a new data structure is difficult, but a few
+common data structures (lists, queues, trees) do have RCU APIs that
+can be used.
+
+RCU splits removal updates to the data structures in two phases:
+
+.. slide:: Removal and Reclamation
+   :inline-contents: True
+   :level: 2
+
+   * **Removal**: removes references to elements. Some old readers may
+     still see the old reference so we can't free the element.
+
+   * **Elimination**: free the element. This action is postponed until
+     all existing readers finish traversal (quiescent cycle). New
+     readers won't affect the elimination delay.
+
+
+As an example, lets take a look on how deleting an element from a list
+using RCU is done:
+
+.. slide:: RCU List Delete
+   :inline-contents: True
+   :level: 2
+
+   |_|
+
+   .. ditaa::
+
+         (1) List Traversal                          (2) Removal
+                                                    +-----------+
+      +-----+     +-----+     +-----+      +-----+  |  +-----+  |  +-----+
+      |     |     |     |     |     |      |     |  |  |     |  |  |     |
+      |  A  |---->|  B  |---->|  C  |      |  A  |--+  |  B  |  +->|  C  |
+      |     |     |     |     |     |      |     |     |     |     |     |
+      +-----+     +-----+     +-----+      +-----+     +-----+     +-----+
+         ^           ^           ^            ^           ^           ^
+         |           |           |            |           |           |
+
+          (3) Quiescent cycle over                 (2) Reclamation
+               +-----------+
+      +-----+  |  +-----+  |  +-----+      +-----+                 +-----+
+      |     |  |  |     |  |  |     |      |     |                 |     |
+      |  A  |--+  |  B  |  +->|  C  |      |  A  |---------------->|  C  |
+      |     |     |     |     |     |      |     |                 |     |
+      +-----+     +-----+     +-----+      +-----+                 +-----+
+         ^                       ^            ^                       ^
+         |                       |            |                       |
+
+
+In the first step it can be seen that while readers traverse the list
+all elements are referenced. In step two a writer removes
+element B. Reclamation is postponed since there are still readers that
+holding references to it. In step three a quiescent cycle just expired
+and it can be noticed that there are no more references to
+element B. Other elements still have references from readers that
+started the list traversal after the element was removed. In step 4 we
+finally perform reclamation (free the element).
+
+
+Now that we covered how RCU functions at the high level, lets looks at
+the APIs for read
+
+
+.. slide:: RCU list APIs cheat sheet
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: c
+
+      /* list traversal */
+      rcu_read_lock();
+      list_for_each_rcu(i, head) {
```

Comment:
There is no longer the `list_for_each_rcu()` macro. We can use the `list_for_each_entry_rcu()` macro: https://elixir.bootlin.com/linux/v4.15/source/include/linux/rculist.h#L341